device_plugins/nvidia_gpu/nvidia_gpu.go

```diff
@@ -0,0 +1,281 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"os/exec"
+	"path"
+	"regexp"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	pluginapi "k8s.io/kubernetes/pkg/kubelet/apis/deviceplugin/v1alpha1"
+)
+
+const (
+	// All NVIDIA GPUs cards should be mounted with nvidiactl and nvidia-uvm
+	// If the driver installed correctly, the 2 devices will be there.
+	nvidiaCtlDevice string = "/dev/nvidiactl"
+	nvidiaUVMDevice string = "/dev/nvidia-uvm"
+	// Optional device.
+	nvidiaUVMToolsDevice string = "/dev/nvidia-uvm-tools"
+	devDirectory                = "/dev"
+	nvidiaDeviceRE              = `^nvidia[0-9]*$`
+
+	// Device plugin settings.
+	devicePluginMountPath = "/device-plugin"
+	kubeletEndpoint       = "kubelet.sock"
+	pluginEndpointPrefix  = "nvidiaGPU"
+	resourceName          = "nvidia.com/GPU"
+)
+
+// nvidiaGPUManager manages nvidia gpu devices.
+type nvidiaGPUManager struct {
+	defaultDevices []string
+	devices        map[string]pluginapi.Device
+	grpcServer     *grpc.Server
+}
+
+func NewNvidiaGPUManager() (*nvidiaGPUManager, error) {
+	return &nvidiaGPUManager{
+		devices: make(map[string]pluginapi.Device),
+	}, nil
+}
+
+// Discovers all NVIDIA GPU devices available on the local node by walking `/dev` directory.
+func (ngm *nvidiaGPUManager) discoverGPUs() error {
+	reg := regexp.MustCompile(nvidiaDeviceRE)
+	files, err := ioutil.ReadDir(devDirectory)
+	if err != nil {
+		return err
+	}
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+		if reg.MatchString(f.Name()) {
+			fmt.Printf("Found Nvidia GPU %q\n", f.Name())
```

Comment:
nit: use glog.

Reply:
I used fm because it is easier to see the output from kubectl logs. Is there a way to get glog output from kubectl logs?